File: backend/src/routes/integrations.ts
import { Router, Request, Response } from 'express';
import { IntegrationModel } from '../models/Integration';
import { ApiResponse } from '../types';

const router = Router();

/**
 * GET /api/integrations - Get all integrations
 */
router.get('/', async (req: Request, res: Response<ApiResponse>) => {
    try {
        const integrations = await IntegrationModel.findAll();
        res.json({ success: true, data: integrations });
    } catch (error: any) {
        console.error('Error fetching integrations:', error);
        res.status(500).json({ success: false, error: error.message });
    }
});

/**
 * GET /api/integrations/:id - Get integration by ID
 */
router.get('/:id', async (req: Request, res: Response<ApiResponse>) => {
    try {
        const integration = await IntegrationModel.findById(req.params.id);
        if (!integration) {
            return res.status(404).json({ success: false, error: 'Integration not found' });
        }
        res.json({ success: true, data: integration });
    } catch (error: any) {
        console.error('Error fetching integration:', error);
        res.status(500).json({ success: false, error: error.message });
    }
});

/**
 * POST /api/integrations - Create new integration
 */
router.post('/', async (req: Request, res: Response<ApiResponse>) => {
    try {
        const integration = await IntegrationModel.create(req.body);
        res.status(201).json({ success: true, data: integration });
    } catch (error: any) {
        console.error('Error creating integration:', error);
        res.status(500).json({ success: false, error: error.message });
    }
});

/**
 * PUT /api/integrations/:id - Update integration
 */
router.put('/:id', async (req: Request, res: Response<ApiResponse>) => {
    try {
        const integration = await IntegrationModel.update(req.params.id, req.body);
        if (!integration) {
            return res.status(404).json({ success: false, error: 'Integration not found' });
        }
        res.json({ success: true, data: integration });
    } catch (error: any) {
        console.error('Error updating integration:', error);
        res.status(500).json({ success: false, error: error.message });
    }
});

/**
 * DELETE /api/integrations/:id - Delete integration
 */
router.delete('/:id', async (req: Request, res: Response<ApiResponse>) => {
    try {
        const success = await IntegrationModel.delete(req.params.id);
        if (!success) {
            return res.status(404).json({ success: false, error: 'Integration not found' });
        }
        res.json({ success: true, message: 'Integration deleted successfully' });
    } catch (error: any) {
        console.error('Error deleting integration:', error);
        res.status(500).json({ success: false, error: error.message });
    }
});

export default router;

File: backend/internal/handlers/integrations.go
package handlers

import (
	"log"
	"net/http"

	"go-backend/internal/models"

	"github.com/gin-gonic/gin"
)

type IntegrationsHandler struct {
	repo *models.IntegrationRepository
}

func NewIntegrationsHandler() *IntegrationsHandler {
	return &IntegrationsHandler{
		repo: models.NewIntegrationRepository(),
	}
}

type CreateIntegrationRequest struct {
	Name     string         `json:"name" binding:"required"`
	Type     string         `json:"type" binding:"required"`
	Status   string         `json:"status"`
	Config   map[string]any `json:"config"`
	LastSync any            `json:"last_sync"`
}

// GetAll returns all integrations
// GET /api/integrations
func (h *IntegrationsHandler) GetAll(c *gin.Context) {
	items, err := h.repo.FindAll()
	if err != nil {
		log.Printf("[IntegrationsHandler] Error fetching integrations: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true, "data": items})
}

// Create creates an integration row
// POST /api/integrations
func (h *IntegrationsHandler) Create(c *gin.Context) {
	var req CreateIntegrationRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"success": false, "error": err.Error()})
		return
	}

	status := models.IntegrationStatusDisconnected
	if req.Status != "" {
		status = models.IntegrationStatus(req.Status)
	}
	if req.Config == nil {
		req.Config = map[string]any{}
	}

	it, err := h.repo.Create(req.Name, req.Type, status, req.Config)
	if err != nil {
		log.Printf("[IntegrationsHandler] Error creating integration: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true, "data": it})
}

// Update updates an integration
// PUT /api/integrations/:id
func (h *IntegrationsHandler) Update(c *gin.Context) {
	id := c.Param("id")

	var updates map[string]any
	if err := c.ShouldBindJSON(&updates); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"success": false, "error": err.Error()})
		return
	}

	it, err := h.repo.Update(id, updates)
	if err != nil {
		log.Printf("[IntegrationsHandler] Error updating integration %s: %v", id, err)
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "error": err.Error()})
		return
	}
	if it == nil {
		c.JSON(http.StatusNotFound, gin.H{"success": false, "error": "Integration not found"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true, "data": it})
}

// Delete removes an integration
// DELETE /api/integrations/:id
func (h *IntegrationsHandler) Delete(c *gin.Context) {
	id := c.Param("id")

	if err := h.repo.Delete(id); err != nil {
		log.Printf("[IntegrationsHandler] Error deleting integration %s: %v", id, err)
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}
